src/reference/mod.rs

```diff
@@ -0,0 +1,404 @@
+//!   (expression_statement [2, 0] - [2, 13]
+//!     (call [2, 0] - [2, 13]
+//!       function: (identifier [2, 0] - [2, 5])
+//!       arguments: (argument_list [2, 5] - [2, 13]
+//!         (identifier [2, 6] - [2, 7])
+//!         (attribute [2, 9] - [2, 12]
+//!           object: (identifier [2, 9] - [2, 10])
+//!           attribute: (identifier [2, 11] - [2, 12])))))
+//!   (print_statement [3, 0] - [3, 13]
+//!     argument: (attribute [3, 6] - [3, 13]
+//!       object: (identifier [3, 6] - [3, 11])
+//!       attribute: (identifier [3, 12] - [3, 13]))))
+//! ```
+//!
+//! There are many interesting use cases where you want to use this parsed syntax tree to create
+//! some _other_ graph-like structure.  This library lets you do that, using a declarative DSL to
+//! identify patterns in the parsed syntax tree, along with rules for which graph nodes and edges
+//! to create for the syntax nodes that match those patterns.
+//!
+//! # Terminology
+//!
+//! One confusing aspect of the graph DSL is that we have to talk about two different kinds of
+//! "node": the nodes of the concrete syntax tree that is our input when executing a graph DSL
+//! file, and the nodes of the graph that we produce as an output.  We will be careful to always
+//! use "syntax node" and "graph node" in this documentation to make it clear which kind of node we
+//! mean.
+//!
+//! # High-level structure
+//!
+//! A graph DSL file consists of one or more **_stanzas_**.  Each stanza starts with a tree-sitter
+//! [**_query pattern_**][query], which identifies a portion of the concrete syntax tree.  The
+//! query pattern should use [**_captures_**][captures] to identify particular syntax nodes in the
+//! matched portion of the tree.  Following the query is a **_block_**, which is a sequence of
+//! statements that construct **_graph nodes_**, link them with **_edges_**, and annotate both with
+//! **_attributes_**.
+//!
+//! [query]: https://tree-sitter.github.io/tree-sitter/using-parsers#pattern-matching-with-queries
+//! [captures]: https://tree-sitter.github.io/tree-sitter/using-parsers#capturing-nodes
+//!
+//! Comments start with a semicolon, and extend to the end of the line.
+//!
+//! Identifiers start with either an ASCII letter or underscore, and all remaining characters are
+//! ASCII letters, numbers, underscores, or hyphens.  (More precisely, they satisfy the regular
+//! expression `/[a-zA-Z_][a-zA-Z0-9_-]*/`.)  Identifiers are used as the names of
+//! [attributes](#attributes), [functions](#functions), and [variables](#variables).
+//!
+//! To execute a graph DSL file against a concrete syntax tree, we execute each stanza in the graph
+//! DSL file _in order_.  For each stanza, we identify each place where the concrete syntax tree
+//! matches the query pattern.  For each of these places, we end up with a different set of syntax
+//! nodes assigned to the query pattern's captures.  We execute the block of statements for each of
```

Comment:
This description of the order in which the stanzas execute might not match the behavior of the system we implemented. Maybe that's intentional though?

In case it's helpful, the thing we originally built (based on the single-pass query execution) processed the matches in the order that they were encountered in the syntax tree during a depth-first traversal. Specifically, a match would be processed as soon as the *last* node in its pattern was found.

Reply:
I think it would be better if the order of execution (between stanzas) was left unspecified. Saying it's in the order in which they appear in the file means you have to keep the entire context in your head when thinking about this (which feels very imperative). This does, however, raise the question of _when_ (or rather _where_) to define the graph nodes that are emitted.

Having the insides of each stanza be executed in order seems fine, however, and indeed required for some of the more complex stanzas (like the one that handles Python modules).

Reply:
> This description of the order in which the stanzas execute might not match the behavior of the system we implemented. Maybe that's intentional though?

Yep, it's intentionally different:

> Saying it's in the order in which they appear in the file means you have to keep the entire context in your head when thinking about this (which feels very imperative). This does, however, raise the question of _when_ (or rather _where_) to define the graph nodes that are emitted.

It actually was originally unspecified, and as Max points out, in the current implementation we throw every stanza's query pattern into a single `TSQuery` instance, which means that the matches for each stanza can be interleaved together.  We've actually found that to be _harder_ to reason about, especially for larger graph DSL files.  I suggested moving to this specified order, where you're _guaranteed_ that earlier stanzas execute _fully_ before later stanzas, because it enables thinking about the stanzas locally, since the execution order depends only on what's in the graph DSL file, and not on the dynamic content of the input file you're processing.  And it's how we were structuring the graph DSL files anyway.

> In case it's helpful, the thing we originally built (based on the single-pass query execution) processed the matches in the order that they were encountered in the syntax tree during a depth-first traversal. Specifically, a match would be processed as soon as the _last_ node in its pattern was found.

I might need to reword how matches _within_ a stanza are ordered based on this.  The proposed new implementation is that each stanza would get its own `TSQuery` instance, and within a stanza we'll get matches however they come out of the query engine.

(@maxbrunsfeld, as an aside, this was the _actual_ prompt for this change! We wanted separate `TSQuery` instances anyway to try to make "match limit exceeded" errors less likely, before realizing that meant we could specify a particular execution order for stanzas.)